pydropsonde/circles.py

```diff
@@ -15,6 +19,146 @@ class Circle:
+                        if ~np.isnan(x)
+                    ]
+                )
+
+        circle_y = np.nanmean(c_yc) / (110.54 * 1000)
+        circle_x = np.nanmean(c_xc) / (111.320 * np.cos(np.radians(circle_y)) * 1000)
+        circle_diameter = np.nanmean(c_r) * 2
+
+        xc = [None] * len(x_coor.T)
+        yc = [None] * len(y_coor.T)
+
+        xc = np.mean(x_coor, axis=0)
+        yc = np.mean(y_coor, axis=0)
+
+        delta_x = x_coor - xc  # *111*1000 # difference of sonde long from mean long
+        delta_y = y_coor - yc  # *111*1000 # difference of sonde lat from mean lat
+
+        new_vars = dict(
+            circle_flight_altitude=self.circle_ds["aircraft_geopotential_altitude"]
+            .mean()
+            .values,
+            circle_time=self.circle_ds["launch_time"].mean().values,
+            circle_lon=circle_x,
+            circle_lat=circle_y,
+            circle_diameter=circle_diameter,
+            dx=(["sonde_id", "alt"], delta_x.values),
+            dy=(["sonde_id", "alt"], delta_y.values),
+        )
+        self.circle_ds = self.circle_ds.assign(new_vars)
+
+        return self
+
+    def fit2d(self, var: str = None):
+        """
+        Estimate a 2D linear model to calculate u-values from x-y coordinates.
+
+        :param var: Variable name to extract from the dataset.
+        :returns: intercept, dudx, dudy, and the original dataset.
+        """
+        u = self.circle_ds[var]
+        # to fix nans, do a copy
+        u_cal = u.copy()
+        # a does not need to be copied as this creates a copy already
+        a = np.stack(
+            [np.ones_like(self.circle_ds.dx), self.circle_ds.dx, self.circle_ds.dy],
+            axis=-1,
+        )
+
+        # for handling missing values, both u and a are set to 0, that way
+        # these items don't influence the fit
+        invalid = (
+            np.isnan(u_cal) | np.isnan(self.circle_ds.dx) | np.isnan(self.circle_ds.dy)
+        )
+        under_constraint = np.sum(~invalid, axis=-1) < 6
+        # Use `.where()` to mask invalid elements with 0
+        u_cal = u_cal.where(~invalid, 0)
+        a[invalid] = 0  # Broadcasting the mask
+
+        a_inv = np.linalg.pinv(a)
+
+        intercept, dudx, dudy = np.einsum("...rm,...m->r...", a_inv, u_cal)
+
+        intercept[under_constraint] = np.nan
+        dudx[under_constraint] = np.nan
+        dudy[under_constraint] = np.nan
+
+        return intercept, dudx, dudy
+
+    def fit2d_xr(
+        self,
+        var: str = None,
+        input_core_dims=["sonde_id"],
+        output_core_dims=["sonde_id"],
+    ):
+        """
+        Apply the fit2d method to the xarray object using xr.apply_ufunc.
+
+        :param var: Variable name for the dataset.
+        :returns: Xarray object with intercept, dudx, dudy, and original dataset.
+        """
+        # Note that we do not pass self.dx and self.dy here; they are used internally by fit2d.
+        return xr.apply_ufunc(
```

Comment:
connected to the dimension comment: I think the way you call xr.apply_ufunc needs to change.

(no guarantee that this works, but something like
```
def fit2d_xr(x, y, u, sonde_dim="sonde_id"):
    # input and output dims must be a list
    return xr.apply_ufunc(
        fit2d,
        x,
        y,
        u,
        input_core_dims=[[sonde_dim], [sonde_dim], [sonde_dim]],
        output_core_dims=[(), (), ()],
    )


def apply_fit2d(circle): 
    for par in tqdm(["u", "v", "q", "ta", "p"]):
        varnames = [par + "0", "d" + par + "dx", "d" + par + "dy"]

        circle = circle.assign(
            dict(
                zip(
                    varnames,
                    fit2d_xr(
                        x=circle.x,
                        y=circle.y,
                        u=circle[par],
                        sonde_dim="sonde_id",
                    ),
                )
            )
        )
    return circle
```
with calls of `get_xy_coords_for_circles` and `apply_fit2d`)